File: SupervisedLearning/FFNNRuby/ai/layer.rb
class Layer
    attr_accessor :neurons

    def initialize(num_neurons:, num_inputs:)
      @neurons = Array.new(num_neurons) {
        Neuron.new(num_inputs)
      }
    end
  
    def activate(inputs)
      @neurons.map { |neuron| neuron.activate(inputs) }
    end

    def calculate_errors(errors, deltas)
      @neurons.each_with_index do |neuron, i|
        neuron.calculate_error(errors, deltas[i])
      end
    end

    def update_weights(inputs, deltas, learning_rate)
      @neurons.each_with_index do |neuron, i|
        neuron.update_weights(inputs, deltas[i], learning_rate)
      end

    end
end

File: SupervisedLearning/classification/catdogs_golang/ai/layer.go
package ai

type Layer struct {
	neurons []*Neuron
}

func NewLayer(numNeurons, numInputs int) *Layer {
	layer := &Layer{
		neurons: make([]*Neuron, numNeurons),
	}
	for i := range layer.neurons {
		layer.neurons[i] = NewNeuron(numInputs)
	}
	return layer
}

func (l *Layer) Activate(inputs []float64) []float64 {
	outputs := make([]float64, len(l.neurons))
	for i, neuron := range l.neurons {
		outputs[i] = neuron.Activate(inputs)
	}
	return outputs
}

func (l *Layer) CalculateErrors(errors *[]float64, delta []float64) {
	for i, neuron := range l.neurons {
		neuron.CalculateError(errors, delta[i])
	}
}

func (l *Layer) UpdateWeights(inputs, delta []float64, learningRate float64) {
	for i, neuron := range l.neurons {
		neuron.UpdateWeights(inputs, delta[i], learningRate)
	}
}
